lib/expand/join.go

```diff
@@ -953,14 +957,44 @@ func (p *Peer) waitForOperation(operator ops.Operator, operation ops.SiteOperati
 				logger.Info("Operation is not ready yet.")
 				continue
 			}
-			log.Info("Operation is ready.")
+			log.Debug("Operation is ready.")
 			return nil
 		case <-p.ctx.Done():
 			return trace.Wrap(p.ctx.Err())
 		}
 	}
 }
 
+func (p *Peer) leave() error {
+	p.Info("Leave cluster.")
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.NodeLeaveTimeout)
+	defer cancel()
+	return p.createShrinkOperation(ctx)
+}
+
+func (p *Peer) createShrinkOperation(ctx context.Context) error {
+	opCtx, err := p.operationContext()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Refresh operation from cluster
+	operation, err := opCtx.Operator.GetSiteOperation(opCtx.Operation.Key())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = opCtx.Operator.CreateSiteShrinkOperation(ctx,
+		ops.CreateSiteShrinkOperationRequest{
+			AccountID:  opCtx.Cluster.AccountID,
+			SiteDomain: opCtx.Cluster.Domain,
+			Servers:    []string{operation.Servers[0].Hostname},
+			Force:      true,
+			// Have cluster avoid triggering remote node cleanup
+			NodeRemoved: true,
```

Comment:
Depending on where exactly join was interrupted, the new node may actually be already online and thus this flag will lead to new serf member staying in the cluster. I remember this b/c we fixed a similar bug with AWS autoscaling a few months ago.

Maybe omit this flag and let shrink follow the normal procedure, i.e. it should figure out itself whether to launch agent or not?

Reply:
I cannot do this without introducing another flag to avoid launching the remote agent (which I understand this flag is already about).
As for your concern - it will not leave the node (i.e. no serf member) as aborting cleans up the services on the node and all of the state (hence the reason why the agent part of the shrink is redundant as it basically does the same).

Reply:
Okay, that makes sense.